Add better performing versions for groupwise and depthwise convolutions

dskhudia opened the pull request:
Summary: Groupwise and depthwise convolutions become faster with this diff

Differential Revision: D16264144

twmht commented:
@dskhudia 

Do you benchmark this PR?

raghuramank100 reviewed (changes requested):
Please add a benchmark test for the kernels and add the benchmarking results before and after the change as part of documentation for this PR. Thanks!

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -64,7 +64,7 @@ def weight(self):
     @weight.setter
     def weight(self, w):
-        self._packed_weight = torch.ops.quantized.fbgemm_conv_prepack(w, self.groups)
+        self._packed_weight = torch.ops.quantized.fbgemm_conv_prepack(w, self.stride, self.padding, self.dilation, self.groups)
```
an unrelated question: do we have unpack for conv?

dskhudia replied:
@jerryzh168 
We made sure to add unpack for conv. Details in https://github.com/pytorch/FBGEMM/pull/103, https://github.com/pytorch/FBGEMM/pull/105, https://github.com/pytorch/FBGEMM/pull/106, https://github.com/pytorch/FBGEMM/commit/815139b1ba72ed43571084f58c7fdf4ca3c991d5, https://github.com/pytorch/FBGEMM/commit/64a2c73a425e4113839f2c2b596ea28d632d20f6

jerryzh168 replied:
is this ready or still ongoing?

jerryzh168 replied:
oh I saw qconv_unpack, Thanks

dskhudia replied:
This is ready. I might have to change your qat modules if they have landed.

dskhudia commented:
> @dskhudia
> 
> Did you benchmark this PR?

@twmht
I'll add some benchmarking results soon. BTW, this is for convolutions on int8 data rather than fp32.

dskhudia commented:
> Please add a benchmark test for the kernels and add the benchmarking results before and after the change as part of documentation for this PR. Thanks!

Sure. I'll add benchmarking results in the summary. I have a followup PR that adds benchmarking. https://github.com/pytorch/pytorch/pull/22895

dskhudia commented:
@raghuramank100 Please see summary for before and after results.

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -17,17 +17,50 @@ namespace {
+    fbgemm::conv_param_t<2> conv_p(
+        1, // Mini-Batch
+        input_channels_per_group * groups, // input channels
+        output_channels,
```
Why do we need mini-batch and image size as parameters to this function if they do not have any effect? Can we remove these arguments?

dskhudia replied:
The interface for packing takes struct conv_param_t as argument. This way we pass the same conv_param_t struct to both conv operation as well as packing. We will have to change the interface and define a new struct if we want to remove these parameters and that would just work for packing only. The other way of doing this is to add a function in fbgemm that would return conv_param_t and does't take mini-batch and imagesize as input parameters.

jianyuh commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -99,11 +99,10 @@ class QConv2dInt8 final : public c10::OperatorKernel {
     PackedConvWeight& pack_ptr =
         cpp_custom_type_hack::cast<PackedConvWeight>(packed_weight);
     auto packB = pack_ptr.w.get();
-    // packB->printPackedMatrix("PackedB inside QConv2dInt8:");
```
Still keep this for debugging purpose? Do we have printPackedMatrix function in the general Conv interface?

dskhudia replied:
We don't. Let me remove this.

dskhudia committed:
Add better performing versions for groupwise and depthwise convolutions (#22869)

Summary:
Groupwise and depthwise convolutions become faster with this diff
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22869

Test Plan:
buck test mode/dev caffe2/test:quantized -- 'test_qconv'  --print-passing-details

```
Running 2 tests
Started new test run: https://our.intern.facebook.com/intern/testinfra/testrun/562950091484224
      ✓ caffe2/test:quantized - test_qconv (test_quantized.TestQuantizedConv) 2.731 1/2 (passed)
Test output:
> test_qconv (test_quantized.TestQuantizedConv) ... ok
>
> ----------------------------------------------------------------------
> Ran 1 test in 2.732s
>
> OK
      ✓ caffe2/test:quantized - test_qconv_unpack (test_quantized.TestQuantizedConv) 5.187 2/2 (passed)
Test output:
> test_qconv_unpack (test_quantized.TestQuantizedConv) ... ok
>
> ----------------------------------------------------------------------
> Ran 1 test in 5.188s
>
> OK
Finished test run: https://our.intern.facebook.com/intern/testinfra/testrun/562950091484224
Summary (total time 15.66s):
  PASS: 2
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0

```

buck test mode/dev caffe2/test:quantized -- 'test_conv_api'
```
Running 2 tests
Started new test run: https://our.intern.facebook.com/intern/testinfra/testrun/3940649676010406
      ✓ caffe2/test:quantized - test_conv_api (test_nn_quantized.ModuleAPITest) 0.040 1/2 (passed)
      ✓ caffe2/test:quantized - test_conv_api (test_quantized_conv.FunctionalAPITest) 5.402 2/2 (passed)
Finished test run: https://our.intern.facebook.com/intern/testinfra/testrun/3940649676010406
Summary (total time 11.83s):
  PASS: 2
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
```

Differential Revision: D16264144

Pulled By: dskhudia

fbshipit-source-id: 959417212f99d3789c2b48361dd36cb864db3594
